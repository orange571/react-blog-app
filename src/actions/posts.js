import uuid from 'uuid';
import database from '../firebase/firebase';
import moment from 'moment'

// ADD_POST
export const addPost = (post) => ({
  type: 'ADD_POST',
  post
});

export const startAddPost = (postData = {}) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    const {
      title = '',
      body = '',
      createdAt = 0,
      isReadable = false,
      author = '',
    } = postData;
    const post = { title, body, createdAt, isReadable, author, uid };

    return database.ref(`posts`).push(post).then((ref) => {
      const postId = ref.key;
      database.ref(`users/${uid}/posts`).update({[postId] : isReadable}).then(()=>{
        if(isReadable){
          database.ref(`isPublic`).push({id: postId})
        };
      });
      dispatch(addPost({
        id: ref.key,
        ...post
      }));
    });
  };
};

// REMOVE_POST
export const removePost = ({ id } = {}) => ({
  type: 'REMOVE_POST',
  id
});

export const startRemovePost = ({ id } = {}) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database.ref(`posts/${id}`).remove().then(() => {
      //REMOVE FROM PUBLIC LIST IF WAS THERE BEFORE
      database.ref(`isPublic`).orderByChild("id").equalTo(id).once("value",snapshot => {
          const wasPublic = snapshot.val();
          if (!!wasPublic){
            const publicPostId = Object.keys(wasPublic)[0];
            database.ref(`isPublic/${publicPostId}`).remove().then(() => {
              //REMOVE FROM USER List
              database.ref(`users/${uid}/posts/${id}`).remove().then(()=>console.log("success remove from user"))
              .catch(error => console.log(error));
            }).catch(error => console.log(error));
          } else {
            //REMOVE FROM USER List
            database.ref(`users/${uid}/posts/${id}`).remove().then(()=>console.log("success remove from user"))
            .catch(error => console.log(error));
          }
      });
      dispatch(removePost({ id }));
    });
  };
};

// EDIT_POST
export const editPost = (id, updates) => ({
  type: 'EDIT_POST',
  id,
  updates
});

export const startEditPost = (id, updates) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database.ref(`posts/${id}`).update(updates).then(() => {
      //ADD TO PUBLIC LIST IF WAS NOT THERE ALREADY
      if(updates.isReadable){
        database.ref(`isPublic`).orderByChild("id").equalTo(id).once("value",snapshot => {
            const wasPublic = snapshot.val();
            if (!wasPublic){
              database.ref(`isPublic`).push({id: id})
            }
        });
        //REMOVE FROM PUBLIC LIST IF WAS THERE BEFORE
      } else {
        database.ref(`isPublic`).orderByChild("id").equalTo(id).once("value",snapshot => {
            const wasPublic = snapshot.val();
            const publicPostId = Object.keys(wasPublic)[0];
            if (!!wasPublic){
              database.ref(`isPublic/${publicPostId}`).remove();
            }
        });
      }
      database.ref(`users/${uid}/posts`).update({[id]: updates.isReadable});
      dispatch(editPost(id, updates));
    });
  };
};

// SET_POSTS
export const setPosts = (posts) => ({
  type: 'SET_POSTS',
  posts
});

export const startSetPosts = () => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database.ref(`users/${uid}/posts`).once('value').then((snapshot) => {
      if(!snapshot.val()) {
        let posts = [];
        dispatch(setPosts(posts));
      } else {
        return new Promise((resolve, reject) => {
          const postIds = [];
          snapshot.forEach((childSnapshot) => {
            postIds.push(childSnapshot.key);
          });
          let posts = [];
          postIds.forEach((postId, index)=>{
            database.ref(`posts/${postId}`).once('value').then((postSnapshot) => {
              posts.push({
                id: postSnapshot.key,
                ...postSnapshot.val()
              });
              if (index === postIds.length-1) {
                resolve(posts);
              }
            }).catch((error) => {
              console.log("error", error);
              reject(error);
            });
          });
        }).then((posts)=>{
          dispatch(setPosts(posts));
        }).catch(error => {
          console.log(error)
        });
      }
    }).catch((error)=>{
      console.log('error', error);
    });
  };
};
