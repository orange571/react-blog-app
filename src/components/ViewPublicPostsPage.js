import React from 'react';
import Header from './Header';
import LoadingPage from './LoadingPage';
import PublicPostsList from './PublicPostsList';
import database from '../firebase/firebase';

class ViewPublicPostsPage extends React.Component {
  constructor (props) {
    super(props);
    this.state = {
      isLoading:false,
      posts: []
    }
  }
  componentDidMount() {
    this.setState((prevState) => {
      return {
        ...prevState,
        isLoading:true
      }
    });
    database.ref('isPublic').once('value').then((snapshot)=> {
      const publicPostIds = snapshot.val();
      const postIds = [];
      snapshot.forEach((childSnapshot)=>{
        postIds.push(childSnapshot.val().id);
      });
      const posts = [];
      if(postIds.length > 0) {
        postIds.forEach((postId, index)=>{
          database.ref(`posts/${postId}`).once('value').then((postSnapshot) => {
            posts.push({
              id: postSnapshot.key,
              ...postSnapshot.val()
            });
            if (index === postIds.length-1) {
              posts.sort((a,b)=> a.createdAt < b.createdAt ? 1 : -1);
              this.setState((prevState) => ({
                ...prevState,
                isLoading:false,
                posts
              }))
            }
          }).catch(error => {
            console.log(error);
            posts.push({
              id: postId,
              title: "error fetching post",
              body: postId
            });
            if (index === postIds.length-1) {
              posts.sort((a,b)=> a.createdAt < b.createdAt ? 1 : -1);
              this.setState((prevState) => ({
                ...prevState,
                isLoading:false,
                posts
              }))
            }
          });
        });
      } else {
        this.setState((prevState) => ({
          ...prevState,
          isLoading:false,
        }))
      }

    }).catch(error => this.setState((prevState) => ({
       ...prevState,
       error,
       isLoading: false
    })));
  }
  render() {
    const isLoading = this.state.isLoading;
    const error = this.state.error;

    if (error) {
      return <p>Something went wrong: {error} </p>
    }

    if (isLoading === true) {
      return <LoadingPage />;
    }
    return (
      <div>
        <Header />
        <PublicPostsList posts={this.state.posts}/>
      </div>
    )
  }
}
export default ViewPublicPostsPage;
