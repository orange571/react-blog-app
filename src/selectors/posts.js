import moment from 'moment';

// Get visible posts

export default (posts, { text, sortBy }) => {
  return posts.filter((post) => {
    const createdAtMoment = moment(post.createdAt);
    const textMatch = post.title.toLowerCase().includes(text.toLowerCase());

    return textMatch;
  }).sort((a, b) => {
    if (sortBy === 'date') {
      return a.createdAt < b.createdAt ? 1 : -1;
    } else if (sortBy === 'title') {
      return a.title.localeCompare(b.title);
    }
  });
};
