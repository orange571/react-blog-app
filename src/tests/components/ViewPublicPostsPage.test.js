import React from 'react';
import configureMockStore from 'redux-mock-store';
import { shallow, mount } from 'enzyme';
import ViewPublicPostsPage from '../../components/ViewPublicPostsPage';
import thunk from 'redux-thunk';
import posts from '../fixtures/posts';
import database from '../../firebase/firebase';

const uid = 'thisismytestuid';
const displayName = "thisismytestname"
const defaultAuthState = { auth: { uid, displayName } };
const createMockStore = configureMockStore([thunk]);

test('should fetch the publicposts from firebase', (done) => {
  const postsData = {};
  posts.forEach(({ id, title, author, body, isReadable, createdAt, uid }) => {
    postsData[id] = { title, author, body, isReadable, createdAt, uid };
  });
  const userPosts = {};
  posts.forEach((post) => {
    userPosts[post.id] = post.isReadable;
  });
  const publicList = {};
  posts.forEach((post) => {
    if(post.isReadable){
      publicList[post.id] = {id: post.id};
    }
  })
  database.ref(`users/${uid}/posts`).set(userPosts).then(() => {
    database.ref(`posts`).set(postsData).then(() => {
      database.ref(`isPublic`).set(publicList).then(() => {
        let wrapper = shallow(<ViewPublicPostsPage />);
        setTimeout(() => {
          expect(wrapper.state().posts).toEqual([posts[1]]);
          expect(wrapper).toMatchSnapshot();
          done();
        }, 1000);
      }).catch(error => console.log(error));
    }).catch(error => console.log(error));
  }).catch(error => console.log(error));

},10000);
