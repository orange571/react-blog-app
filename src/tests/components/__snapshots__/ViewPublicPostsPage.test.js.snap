// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should fetch the publicposts from firebase 1`] = `<LoadingPage />`;
